### PR TITLE
Cb 4268 force logout

### PR DESCRIPTION
**Pull request by DenisSinelnikov**

### CODE REVIEW

File: plugins/org.jkiss.dbeaver.model.sm/src/org/jkiss/dbeaver/model/auth/SMAuthInfo.java [hunk 1]
```diff
@@ -22,7 +22,7 @@
 
 import java.util.Map;
 
-public class SMAuthInfo {
+public class SMAuthInfo { //create flag
```
Comment (alexander-skoblikov) [inline plugins/org.jkiss.dbeaver.model.sm/src/org/jkiss/dbeaver/model/auth/SMAuthInfo.java]: remove comment

File: plugins/org.jkiss.dbeaver.model.sm/src/org/jkiss/dbeaver/model/security/exception/SMExceptionTooManySessions.java [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jkiss.dbeaver.model.security.exception;
+
+public class SMExceptionTooManySessions extends SMException {
```
Comment (alexander-skoblikov) [inline plugins/org.jkiss.dbeaver.model.sm/src/org/jkiss/dbeaver/model/security/exception/SMExceptionTooManySessions.java]:
```suggestion
public class SMTooManySessionsException extends SMException {
```

File: plugins/org.jkiss.dbeaver.model.event/src/org/jkiss/dbeaver/model/websocket/event/WSUserEvent.java [hunk 3]
```diff
@@ -18,18 +18,24 @@
 
 import org.jkiss.code.NotNull;
 
-public class WSUserDeletedEvent extends WSAbstractEvent {
+import java.util.List;
+
+public class WSUserEvent extends WSAbstractEvent {
```
Comment (alexander-skoblikov) [inline plugins/org.jkiss.dbeaver.model.event/src/org/jkiss/dbeaver/model/websocket/event/WSUserEvent.java]: Create separate event for force closed sessions, do not mix it with user delete event

**Review by alexander-skoblikov — CHANGES_REQUESTED** [review]

File: plugins/org.jkiss.dbeaver.model.sm/src/org/jkiss/dbeaver/model/security/exception/SMExceptionTooManySessions.java [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+
+    private final String errorType;
+
+    public SMExceptionTooManySessions(String message, String errorType) {
```
Comment (alexander-skoblikov) [inline plugins/org.jkiss.dbeaver.model.sm/src/org/jkiss/dbeaver/model/security/exception/SMExceptionTooManySessions.java]: do we really need `errorType` as parameter? 
it not a general error, and it seems we can just hardcode `tooManySessions` here

File: plugins/org.jkiss.dbeaver.model.event/src/org/jkiss/dbeaver/model/websocket/event/WSUserEvent.java [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+package org.jkiss.dbeaver.model.websocket.event;
+
+import org.jkiss.code.NotNull;
+
+import java.util.List;
+
+public class WSUserEvent extends WSAbstractEvent {
+
+    @NotNull
+    private final List<String> ids;
```
Comment (alexander-skoblikov) [inline plugins/org.jkiss.dbeaver.model.event/src/org/jkiss/dbeaver/model/websocket/event/WSUserEvent.java]:
```suggestion
    private final List<String> sessionIds;
```

Comment (alexander-skoblikov) [reply plugins/org.jkiss.dbeaver.model.event/src/org/jkiss/dbeaver/model/websocket/event/WSUserEvent.java]: move this fied to WsUserCloseSessionsEvent

File: plugins/org.jkiss.dbeaver.model.sm/src/org/jkiss/dbeaver/model/auth/SMAuthInfo.java [hunk 6]
```diff
@@ -56,6 +56,9 @@ public class SMAuthInfo {
     private final SMAuthPermissions authPermissions;
 
     private final boolean mainAuth;
+    private final boolean forceSessionsLogout;
+
+    private final String errorCode;
```
Comment (alexander-skoblikov) [inline plugins/org.jkiss.dbeaver.model.sm/src/org/jkiss/dbeaver/model/auth/SMAuthInfo.java:62]: nullable annotation

File: plugins/org.jkiss.dbeaver.model.sm/src/org/jkiss/dbeaver/model/security/exception/SMTooManySessionsException.java [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+
+public class SMTooManySessionsException extends SMException {
+
+    private final String errorType = "tooManySessions";;
+
+    public SMTooManySessionsException(String message) {
+        super(message);
+    }
+
+    public String getErrorType() {
```
Comment (alexander-skoblikov) [inline plugins/org.jkiss.dbeaver.model.sm/src/org/jkiss/dbeaver/model/security/exception/SMTooManySessionsException.java:30]: nullable annotation

**Review by alexander-skoblikov — CHANGES_REQUESTED** [review]

File: plugins/org.jkiss.dbeaver.model.sm/src/org/jkiss/dbeaver/model/auth/SMAuthenticationManager.java [hunk 8]
```diff
@@ -27,8 +27,9 @@ void updateAuthStatus(
         @NotNull String authId,
         @NotNull SMAuthStatus authStatus,
         @NotNull Map<SMAuthConfigurationReference, Object> authInfo,
-        @Nullable String error
+        @Nullable String error,
+        @Nullable String errorCode
     ) throws DBException;
 
-    SMAuthInfo finishAuthentication(@NotNull String authId) throws DBException;
+    SMAuthInfo finishAuthentication(@NotNull String authId, boolean forceSessionsLogout) throws DBException;
```
Comment (alexander-skoblikov) [inline plugins/org.jkiss.dbeaver.model.sm/src/org/jkiss/dbeaver/model/auth/SMAuthenticationManager.java]:
```suggestion
    SMAuthInfo finishAuthentication(@NotNull String authId) throws DBException;
```

File: plugins/org.jkiss.dbeaver.model.sm/src/org/jkiss/dbeaver/model/auth/SMAuthInfo.java [hunk 9]
```diff
@@ -292,6 +315,16 @@ public Builder setMainAuth(boolean mainAuth) {
             return this;
         }
 
+        public Builder serForceSessionsLogout(boolean forceSessionsLogout) {
```
Comment (yagudin10) [inline plugins/org.jkiss.dbeaver.model.sm/src/org/jkiss/dbeaver/model/auth/SMAuthInfo.java]:
```suggestion
        public Builder setForceSessionsLogout(boolean forceSessionsLogout) {
```

**Review by alexander-skoblikov — APPROVED** [review]

**Pull request merged**